File: Caro_pybind/Caro_pybind/mcts.cpp
#include "mcts.h"


void MCTS_AI::expand_node(TreeNode *node)
{
    expanded_nodes_count++;
    unsigned int depth = node->turn_count + 1;
    if (depth > current_max_depth)
    {
        current_max_depth = depth;
    }
    std::set<Point> moves = board.get_AI_moves();
    auto board_temp = board.get_board();
    for (Point const& p : moves)
    {
        auto* child = new TreeNode(p, -node->player, node, depth);
        nodes_vector.push_back(child);
        node->children.push_back(child);

        if (use_prior)  // evaluate prior for each node if a prior evaluator is given
        {
            board_temp[p(0)][p(1)] = child->player;
            float prior;
            if (model != nullptr)
            {
                prior = eval_prior_model(board_temp, board.get_dim());
            }
            else if (evaluate_prior != nullptr)
            {
                prior = evaluate_prior(board_temp, board.get_dim());
            }
            else
            {
                std::cout << "NO MODEL TO EVALUATE PRIOR\n";
            }
            child->prior_eval = prior;
            board_temp[p(0)][p(1)] = 0;
        }
    }
    child_count += node->children.size();
}

float MCTS_AI::posterior_eval(TreeNode* node) const
{
    if (prior_strength == 0 && node->visit_count == 0)
    {
        return 0;
    }
    return node->player * (node->prior_eval * prior_strength + node->total_reward) / (prior_strength + node->visit_count);
}

float MCTS_AI::evaluate_uct(TreeNode* node) const
{
    if (!use_prior)
    {
        return node->uct();
    }
    return ((posterior_eval(node) + 1) * 0.5 + 1) + node->exploration_value(use_prior);     // shift reward to [1, 2] to make sure UCT > 0
}

float MCTS_AI::mcts(TreeNode* node, bool weighted_select)
{
    if (node->visit_count >= min_visits)    // matured node
    {
        node->visit_count++;
        if (node->children.empty())     // Initialize child nodes if empty
        {
            expand_node(node);
        }
        TreeNode* next;
        if (weighted_select)
        {
            next = mcts_selection_temperature(node);
        }
        else
        {
            next = mcts_selection(node);
        }
        board.play(next->move);
        if (board.has_ended())     // If the game ends, accumulate reward for next and current node, then propagate result back up
        {
            next->visit_count++;
            int result = board.get_state();
            next->total_reward += result;
            node->total_reward += result;
            board.undo();
            return result;
        }
        float result = mcts(next); // recursion call mcts on node next if game doesn't end (no weighted selection)
        board.undo();
        // propagate result back up
        node->total_reward += result;
        return result;
    }
    else        // not enough maturity
    {
        node->visit_count++;
        float result;
        if (use_prior)
        {
            if (rollout_weight < 0.001f)
            {
                result = node->prior_eval;
            }
            else
            {
                result = rollout_weight * simulate() + (1 - rollout_weight) * node->prior_eval;
            }
        }
        else
        {
            result = simulate();
        }
        if (node->visit_count >= min_visits)    // if matured
        {
            expand_node(node);
        }
        // propagate result back up
        node->total_reward += result;
        return result;
    }
}

TreeNode* MCTS_AI::mcts_selection(TreeNode *node)
{
    TreeNode* current = node->children[0];
    double current_eval = evaluate_uct(current);
    double child_eval = 0;
    for (TreeNode* child : node->children)
    {
        child_eval = evaluate_uct(child);
        if (current_eval < child_eval)
        {
            current = child;
            current_eval = child_eval;
        }
    }
    return current;
}

void MCTS_AI::evaluate_children_weights(TreeNode* node)
{
    if (uct_temperature < 0.001)
    {
        return;
    }

    float exponent = 1 / uct_temperature;

    for (TreeNode* child : node->children)
    {
        child->weight = pow(evaluate_uct(child), exponent);
    }
}

// MCTS select according to weighted UCT score
TreeNode* MCTS_AI::mcts_selection_temperature(TreeNode* node)
{
    if (uct_temperature < 0.01)
    {
        return mcts_selection(node);
    }

    evaluate_children_weights(node);
    float total_weights = 0;
    for (int i = 0; i < node->children.size(); i++)
    {
        total_weights += node->children[i]->weight;
    }

    std::mt19937 rng(std::random_device{}());
    std::uniform_real_distribution<float> udis(0, total_weights);
    float rnd = udis(rng);

    int temp = 0;
    for (int i = 0; i < node->children.size(); i++)
    {
        temp = i;
        if (rnd <= node->children[i]->weight)
        {
            return node->children[i];
        }
        rnd -= node->children[i]->weight;
    }

    std::cout << "RANDOM SELECTION DID NOT PICK ANY\n";
    return  node->children[temp];
}

int MCTS_AI::simulate()
{
    int current_turn = board.get_turn_count();
    board.simulate();
    int end_turn = board.get_turn_count();
    int final_state = board.get_state();
    for (int i = 0; i < end_turn - current_turn; i++)
    {
        board.undo();
    }
    return final_state;
}

// Return the move the AI think is best according to mode
Point MCTS_AI::get_move(Point prev_move)
{
    if (prev_move != Point(-1, -1))
    {
        if (current_node == nullptr || !(prev_move == current_node->move && player != current_node->player))    // in the case of not playing against itself
        {
            bool valid = board.play(prev_move);
            if (!valid)
            {
                std::cout << "INVALID MOVE " << prev_move.to_string() << "\n";
                return Point(-1, -1);
            }
        }
    }
    // AI first move, current_node will be nullptr, current_node is the previous move of the other player
    if (current_node == nullptr)
    {
        current_node = new TreeNode(prev_move, -player, nullptr, board.get_turn_count());
        nodes_vector.push_back(current_node);
        expand_node(current_node);
    }
    else    // update current_node to be its child with prev_move
    {
        bool found_move = false;
        for (TreeNode* child : current_node->children)
        {
            if (child->move == prev_move)
            {
                current_node = child;
                found_move = true;
                break;
            }
        }
        if (!found_move)    // There is no child node with prev_move
        {
            current_node = new TreeNode(prev_move, -player, current_node, board.get_turn_count());
            nodes_vector.push_back(current_node);
        }
    }

    if (current_node->children.empty())     // Initialize child nodes if empty
    {
        expand_node(current_node);
    }

    current_depth = current_node->turn_count;

    // MCTS for n_sim iterations
    for (int n = 0; n < n_sim; n++)
    {
        mcts(current_node, true);   // weighted selection at root_node only
    }

    TreeNode* next_move;

    if (mode == "random")
    {
        next_move = random_selection(current_node);
    }
    else if (mode == "mcts")
    {
        next_move = mcts_selection(current_node);
    }
    else if (mode == "greedy_post")
    {
        next_move = posterior_selection(current_node);
    }
    else if (mode == "greedy")
    {
        next_move = reward_selection(current_node);
    }
    else if (mode == "greedy_visit")
    {
        next_move = visit_selection(current_node);
    }
    else if (mode == "weighted_visit")
    {
        next_move = weighted_visit_selection(current_node);
    }
    else if (mode == "weighted_reward")
    {
        next_move = weighted_reward_selection(current_node);
    }
    else if (mode == "alpha_zero")
    {
        if (board.get_turn_count() <= random_threshold)
        {
            next_move = weighted_visit_selection(current_node);
        }
        else
        {
            next_move = visit_selection(current_node);
        }
    }
    else if (mode == "alpha_zero_reward")
    {
        if (board.get_turn_count() <= random_threshold)
        {
            next_move = weighted_reward_selection(current_node);
        }
        else
        {
            next_move = reward_selection(current_node);
        }
    }
    else
    {
        std::cout << "INVALID MODE" << std::endl;
        exit(69);
    }

    return next_move->move;
}

// Tell the AI to play a move on its board
void MCTS_AI::play_move(Point move)
{
    bool valid = board.play(move);
    if (!valid)
    {
        std::cout << "INVALID MOVE " << move.to_string() << "\n";
        return;
    }

    bool found_move = false;
    for (TreeNode* child : current_node->children)
    {
        if (child->move == move)
        {
            current_node = child;
            found_move = true;
            break;
        }
    }

    if (!found_move)    // There is no child node with move
    {
        current_node = new TreeNode(move, player, current_node, board.get_turn_count());
        nodes_vector.push_back(current_node);
    }

    if (current_node->children.empty())     // Initialize child nodes if empty
    {
        expand_node(current_node);
    }

}

// FOR DIM=7 only
std::array<int, 49> MCTS_AI::get_search_distribution()
{
    std::array<int, 49> search_vector{};
    for (TreeNode* child : current_node->children)
    {
        search_vector[child->move(0) * 7 + child->move(1)] = child->visit_count;
    }
    return search_vector;
}

// Pick move based on average_reward
TreeNode *MCTS_AI::reward_selection(TreeNode *node) {
    TreeNode* current = node->children[0];      // THIS LINE IS BUGGED FOR SOME REASON AFRER USING GET CURRENT_NODE IN PYTHON -> BECAUSE GARBAGE COLLECTOR OMEGALUL
    for (TreeNode* child : node->children)
    {
        if (current->average_reward() < child->average_reward())
        {
            current = child;
        }
    }
    return current;
}

TreeNode* MCTS_AI::random_selection(TreeNode* node)
{
    int rand_index = std::rand() % node->children.size();
    return node->children[rand_index];
}

TreeNode* MCTS_AI::visit_selection(TreeNode* node)
{
    TreeNode* current = node->children[0];
    for (TreeNode* child : node->children)
    {
        if (current->visit_count < child->visit_count)
        {
            current = child;
        }
    }
    return current;
}

// Select node proportional to its visit count
TreeNode* MCTS_AI::weighted_visit_selection(TreeNode* node)
{
    if (play_temperature < 0.001)
    {
        return visit_selection(node);
    }

    float total_weights = 0;
    std::vector<float> weight_vector;
    float exponent = 1 / play_temperature;
    for (int i = 0; i < node->children.size(); i++)
    {
        float w = pow(node->children[i]->visit_count, exponent);
        total_weights += w;
        weight_vector.push_back(w);
    }

    std::mt19937 rng(std::random_device{}());
    std::uniform_real_distribution<float> udis(0, total_weights);
    float rnd = udis(rng);
    int temp = 0;
    for (int i = 0; i < node->children.size(); i++)
    {
        temp = i;
        if (rnd <= weight_vector[i])
        {
            return node->children[i];
        }
        rnd -= weight_vector[i];
    }

    std::cout << "RANDOM SELECTION DID NOT PICK ANY\n";
    return  node->children[temp];
}

TreeNode* MCTS_AI::weighted_reward_selection(TreeNode* node)
{
    if (play_temperature < 0.001)
    {
        return reward_selection(node);
    }

    float total_weights = 0;
    std::vector<float> weight_vector;
    float exponent = 1 / play_temperature;
    for (int i = 0; i < node->children.size(); i++)
    {
        float shifted_reward = (node->children[i]->average_reward() + 1) * 0.5 + 1;     // shift reward to be in [1, 2]
        float w = pow(shifted_reward, exponent);
        total_weights += w;
        weight_vector.push_back(w);
    }

    std::mt19937 rng(std::random_device{}());
    std::uniform_real_distribution<float> udis(0, total_weights);
    float rnd = udis(rng);
    int temp = 0;
    for (int i = 0; i < node->children.size(); i++)
    {
        temp = i;
        if (rnd <= weight_vector[i])
        {
            return node->children[i];
        }
        rnd -= weight_vector[i];
    }

    std::cout << "RANDOM SELECTION DID NOT PICK ANY\n";
    return  node->children[temp];
}

TreeNode* MCTS_AI::posterior_selection(TreeNode* node)
{
    TreeNode* current = node->children[0];
    float current_eval = posterior_eval(current);
    float child_eval = 0;
    for (TreeNode* child : node->children)
    {
        child_eval = posterior_eval(child);
        if (current_eval < child_eval)
        {
            current = child;
            current_eval = child_eval;
        }
    }
    return current;
}


MCTS_AI::~MCTS_AI()
{
    for (TreeNode* node : nodes_vector)
    {
        delete node;
    }
}

File: MCTS.py
from Caro_pybind import Caro, Point
from typing import Callable
import math
import random
from copy import copy


CHAR_P = {-1: "O", 0: ".", 1: "X"}
C = math.sqrt(2)


class TreeNode:
    def __init__(self, move: Point, player: int, parent=None, turn_count: int = 0):
        self.move = move
        self.player = player
        self.parent: TreeNode = parent
        self.children: list[TreeNode] = list()

        self.visit_count = 0
        self.total_reward = 0
        self.prior_eval = 0
        self.noise = 0

        self.turn_count = 0
        if turn_count > 0:
            self.turn_count = turn_count
        elif self.parent is not None:
            self.turn_count = self.parent.turn_count + 1

    def __str__(self):
        return "TreeNode: " + str(self.move) + " player: " + CHAR_P[self.player]

    def average_reward(self) -> float:
        if self.visit_count == 0:
            return 0
        return self.player * self.total_reward / self.visit_count

    def exploration_value(self, use_prior: bool = False) -> float:
        if self.visit_count + int(use_prior) == 0:
            return 9999
        return C * math.sqrt(math.log(self.parent.visit_count) / (self.visit_count + int(use_prior)))

    def uct(self) -> float:
        return self.average_reward() + self.exploration_value()


class MCTS:
    def __init__(self, _player: int, _min_visits: int, _n_sim: int, _board: Caro, _mode: str,
                 _ai_moves_range: int = 1, _eval: Callable = None,
                 _prior_strength: int = 1, _random_threshold: int = 8):
        self.player = _player
        self.min_visits = _min_visits
        self.n_sim = _n_sim
        self.board: Caro = copy(_board)

        self.current_node: TreeNode = None
        self.AI_moves_range = _ai_moves_range
        self.mode = _mode
        self.random_threshold = _random_threshold

        self.use_prior = False
        self.evaluate_prior = _eval
        self.prior_strength = 1
        if self.evaluate_prior is not None:
            self.use_prior = True
            if _prior_strength >= 0:
                self.prior_strength = _prior_strength

        self.child_count = 0
        self.expanded_nodes_count = 0
        self.current_depth = 0
        self.current_max_depth = 0

    def expand_node(self, node: TreeNode) -> None:
        self.expanded_nodes_count += 1
        depth = node.turn_count + 1
        if depth > self.current_max_depth:
            self.current_max_depth = depth

        moves = self.board.get_AI_moves()
        board_temp = self.board.get_board()
        for p in moves:
            child = TreeNode(p, -node.player, node, depth)
            node.children.append(child)
            # Evaluate prior for each node if a prior is given
            if self.use_prior:
                board_temp[p(0)][p(1)] = child.player
                child.prior_eval = float(self.evaluate_prior(board_temp, self.board.get_dim()))
                board_temp[p(0)][p(1)] = 0

        self.child_count += len(node.children)

    def posterior_eval(self, node: TreeNode) -> float:
        if self.prior_strength == 0 and node.visit_count == 0:
            return 0
        return node.player * (node.prior_eval * self.prior_strength + node.total_reward) / (self.prior_strength + node.visit_count)

    def evaluate_node(self, node: TreeNode) -> float:
        if self.use_prior:
            return self.posterior_eval(node) + node.exploration_value(self.use_prior)
        return node.uct()

    def mcts(self, node: TreeNode) -> float:
        if node.visit_count >= self.min_visits:     # Matured Node
            node.visit_count += 1
            if len(node.children) == 0:             # Initialize children if empty
                self.expand_node(node)
            next_node = self.mcts_selection(node)
            self.board.play(next_node.move)
            if self.board.has_ended():              # If the game ends, accumulate reward for next and current node
                next_node.visit_count += 1
                result = self.board.get_state()     # backprop result
                next_node.total_reward += result
                node.total_reward += result
                self.board.undo()
                return result
            result = self.mcts(next_node)           # recursion call mcts on node next if game doesn't end
            self.board.undo()
            # backprop result back up
            node.total_reward += result
            return result
        else:   # not enough maturity
            node.visit_count += 1
            if self.use_prior:
                result = node.prior_eval
            else:
                result = self.simulate()
            if node.visit_count >= self.min_visits:  # if matured
                self.expand_node(node)
            # backprop result back up
            node.total_reward += result
            return result

    def simulate(self) -> int:
        current_turn = self.board.get_turn_count()
        self.board.simulate()
        end_turn = self.board.get_turn_count()
        final_state = self.board.get_state()
        for i in range(end_turn - current_turn):
            self.board.undo()
        return final_state

    def mcts_selection(self, node: TreeNode) -> TreeNode:
        current = node.children[0]
        current_eval = self.evaluate_node(current)
        for child in node.children:
            child_eval = self.evaluate_node(child)
            if current_eval < child_eval:
                current = child
                current_eval = child_eval
        return current

    def posterior_selection(self, node: TreeNode) -> TreeNode:
        current = node.children[0]
        for child in node.children:
            if self.posterior_eval(current) < self.posterior_eval(child):
                current = child
        return current

    @staticmethod
    def reward_selection(node: TreeNode) -> TreeNode:
        current = node.children[0]
        for child in node.children:
            if current.average_reward() < child.average_reward():
                current = child
        return current

    @staticmethod
    def random_selection(node: TreeNode) -> TreeNode:
        rand_index = random.randrange(len(node.children))
        return node.children[rand_index]

    @staticmethod
    def visit_selection(node: TreeNode) -> TreeNode:
        current = node.children[0]
        for child in node.children:
            if current.visit_count < child.visit_count:
                current = child
        return current

    @staticmethod
    def weighted_visit_selection(node: TreeNode) -> TreeNode:
        weights = list(child.visit_count for child in node.children)
        return random.choices(node.children, weights=weights)[0]

    def get_move(self, prev_move: Point) -> Point:
        if prev_move != Point(-1, -1):
            self.board.play(prev_move)
        # AI first move, current_node will be None
        if self.current_node is None:
            self.current_node = TreeNode(prev_move, -self.player, None, self.board.get_turn_count())
            self.expand_node(self.current_node)
        else:   # update current_node to be its child with prev_move
            for child in self.current_node.children:
                if child.move == prev_move:
                    self.current_node = child
                    break
        # Initialize child nodes if empty
        if len(self.current_node.children) == 0:
            self.expand_node(self.current_node)

        self.current_depth = self.current_node.turn_count
        # MCTS for self.n_sim iterations
        for n in range(self.n_sim):
            self.mcts(self.current_node)

        if self.mode == "random":
            self.current_node = self.random_selection(self.current_node)
        elif self.mode == "greedy_post":
            self.current_node = self.posterior_selection(self.current_node)
        elif self.mode == "greedy":
            self.current_node = self.reward_selection(self.current_node)
        elif self.mode == "greedy_visit":
            self.current_node = self.visit_selection(self.current_node)
        elif self.mode == "weighted_visit":
            self.current_node = self.weighted_visit_selection(self.current_node)
        elif self.mode == "alpha_zero":
            if self.board.get_turn_count() <= self.random_threshold:
                self.current_node = self.weighted_visit_selection(self.current_node)
            else:
                self.current_node = self.visit_selection(self.current_node)
        else:
            print("INVALID MODE")
            raise ValueError("INVALID MODE REEEEEEEEEEEE")
        self.board.play(self.current_node.move)
        # Initialize child nodes if empty
        if len(self.current_node.children) == 0:
            self.expand_node(self.current_node)
        self.current_node.parent = None     # Garbage collector the rest of the trees
        return self.current_node.move

    def get_tree_depth(self) -> int:
        return self.current_max_depth - self.current_depth

    def average_child_count(self) -> float:
        if self.expanded_nodes_count == 0:
            return 0
        return self.child_count/self.expanded_nodes_count

    def get_player(self) -> int:
        return self.player

    def predicted_reward(self) -> float:
        if self.mode == "greedy_post":
            return self.player * self.posterior_eval(self.current_node)
        return self.player * self.current_node.average_reward()
